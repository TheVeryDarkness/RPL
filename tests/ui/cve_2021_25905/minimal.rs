//@ revisions: inline regular
//@[inline] compile-flags: -Z inline-mir=true
//@[regular] compile-flags: -Z inline-mir=false

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn main() {
    let mut buf: Vec<u8> = Vec::new();
    let b = buf.len();
    let buf = unsafe {
        std::slice::from_raw_parts_mut(buf.as_mut_ptr().offset(b as isize), buf.capacity() - b)
        //~^ ERROR: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
    };
}
