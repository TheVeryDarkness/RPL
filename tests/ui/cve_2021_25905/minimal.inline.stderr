error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2021_25905/minimal.rs:20:38
   |
LL |                     buf.as_mut_ptr().offset(b as isize),
   |                         ------------ ^^^^^^^^^^^^^^^^^^ offset here
   |                         |
   |                         pointer used here
...
LL |     cases!(Vec::new());
   |     ------------------ in this macro invocation
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: `-D rpl::unchecked-pointer-offset` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::unchecked_pointer_offset)]`
   = note: this error originates in the macro `cases` (in Nightly builds, run with -Z macro-backtrace for more info)

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2021_25905/minimal.rs:20:38
   |
LL |                     buf.as_mut_ptr().offset(b as isize),
   |                         ------------ ^^^^^^^^^^^^^^^^^^ offset here
   |                         |
   |                         pointer used here
...
LL |     cases!(vec![1, 2, 3]);
   |     --------------------- in this macro invocation
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: this error originates in the macro `cases` (in Nightly builds, run with -Z macro-backtrace for more info)

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2021_25905/minimal.rs:20:38
   |
LL |                     buf.as_mut_ptr().offset(b as isize),
   |                         ------------ ^^^^^^^^^^^^^^^^^^ offset here
   |                         |
   |                         pointer used here
...
LL |     cases!(Vec::with_capacity(0));
   |     ----------------------------- in this macro invocation
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: this error originates in the macro `cases` (in Nightly builds, run with -Z macro-backtrace for more info)

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2021_25905/minimal.rs:20:38
   |
LL |                     buf.as_mut_ptr().offset(b as isize),
   |                         ------------ ^^^^^^^^^^^^^^^^^^ offset here
   |                         |
   |                         pointer used here
...
LL |     cases!(Vec::with_capacity(1));
   |     ----------------------------- in this macro invocation
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: this error originates in the macro `cases` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 4 previous errors

